### PR TITLE
Support more keras layers

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

1. Add max, mul, dot to merge layers
2. Add advanced activation layers
3. Add Zeropadding layers
4. Add Cropping1D layer
5. Add recurrent layers
6. Add timedistributeddense layer
7. Add repeatvector layer
8. Add unit tests for each additional layer
9. Add a test for vgg16
10. Other fix in style and others.

## How was this patch tested?

Unit Test.

### CODE REVIEW

File: pyspark/bigdl/keras1/converter.py [hunk 1]
```diff
@@ -751,7 +956,7 @@ def create_globalaveragepooling1d(self, klayer, kclayer):
         return seq
 
     def combo_parameter_layer(self, blayer, config):
-        if config["W_constraint"] or config["b_constraint"]:
+        if "W_constraint" in config or "b_constraint" in config:
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py]: revert this back?

File: pyspark/bigdl/keras1/converter.py [hunk 2]
```diff
@@ -468,7 +675,7 @@ def to_bigdl_2d_padding(self, border_mode):
     def to_bigdl_1d_padding(self, border_mode, kernel_w):
         if border_mode == "same":
             raise Exception("We don't support padding for now")
-            return int((kernel_w -1) / 2)
+            # return int((kernel_w -1) / 2)
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py]: keep this line and add a TODO here

File: pyspark/bigdl/keras1/converter.py [hunk 3]
```diff
@@ -508,7 +715,7 @@ def create_convolution1d(self, klayer, kclayer):
         bpadW, bpadH = self.to_bigdl_2d_padding(klayer.border_mode)
         seq = BLayer.Sequential()
         seq.add(BLayer.View([input_shape[1], 1, input_shape[2]], num_input_dims=3))
-        raise Exception("What !!!!")
+        # raise Exception("What !!!!")
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py]: remove this line

File: pyspark/test/bigdl/keras1/test_layer.py [hunk 4]
```diff
@@ -267,6 +268,120 @@ def test_merge_concat(self):
...
+
+    def test_cropping1d(self):
+        input_data = np.random.uniform(0, 1, [3, 4, 3])
+        layer = Cropping1D(cropping=(1, 2), input_shape=(4, 3))
+        self.__modelTestSingleLayer(input_data, layer, functional_api=False)
+        self.__modelTestSingleLayer(input_data, layer, functional_api=True)
+
+    def test_simplernn(self):
+        input_data = np.random.random([3, 4, 5])
+        layer = SimpleRNN(5, input_shape=(4, 5), return_sequences=True)
```
Comment (zhichao-li) [inline pyspark/test/bigdl/keras1/test_layer.py]: add more test for return_sequences=False

File: pyspark/test/bigdl/keras1/test_layer.py [hunk 5]
```diff
@@ -55,7 +53,11 @@ def __dump_keras(self, keras_model, dump_weights=False):
         return keras_model_json_path, keras_model_hdf5_path
 
     def __generate_model(self, input_data, output_layer):
-        input_node = [Input(shape=input_data.shape[1:]) for i in input_data] # it's a list incase of Merge Layer
+        import keras
```
Comment (zhichao-li) [inline pyspark/test/bigdl/keras1/test_layer.py]: move `import keras` to the header of this file

File: pyspark/test/bigdl/keras1/test_layer.py [hunk 6]
```diff
@@ -55,7 +53,11 @@ def __dump_keras(self, keras_model, dump_weights=False):
...
+        if isinstance(output_layer, keras.engine.Merge):  # it's a list in case of Merge Layer
+            input_node = [Input(shape=i.shape[1:]) for i in input_data]
+        else:
+            input_node = Input(shape=input_data.shape[1:])
```
Comment (zhichao-li) [inline pyspark/test/bigdl/keras1/test_layer.py]:
``` python
          def without_batch(batch_shape):
                   return batch_shape.shape[1:]
+        if isinstance(output_layer, keras.engine.Merge):  # it's a list in case of Merge Layer
              assert isinstance(List, input_data)
+            input = [Input(shape=without_batch(input_data.shape)) for i in input_data]
+        else:
+            input = Input(shape=without_batch(input_data.shape))
```

File: pyspark/test/bigdl/keras1/test_layer.py [hunk 7]
```diff
@@ -86,25 +88,19 @@ def __modelTestSingleLayer(self,
...
 
-    def bigdl_assert_allclose(self, a, b, rtol=1e-7):
-        if a.shape != b.shape and a.shape[0] == 1:
-            a = a.squeeze(0) # bigdl has a leading 1 for conv2d
-            b = b
-        if a.shape != b.shape:
-            a = a.transpose() # for Dense in keras and linear in bigdl has diff order
-        assert_allclose(a, b, rtol)
+    def bigdl_assert_allclose(self, klayer, bweights, kweights, rtol=1e-7):
+        bweights2 = WeightsConverter.convert_weights(klayer.__class__.__name__, kweights)
```
Comment (zhichao-li) [inline pyspark/test/bigdl/keras1/test_layer.py]: bweights_from_keras

File: pyspark/bigdl/keras1/converter.py [hunk 8]
```diff
@@ -403,16 +408,218 @@ def create_reshape(self, klayer, kclayer):
...
+        elif klayer.mode == "sum":
+            blayer = BLayer.CAddTable(
+                inplace=False,
+                bigdl_type="float")
+        elif klayer.mode == "mul":
+            blayer = BLayer.CMulTable(bigdl_type="float")
+        elif klayer.mode == "max":
+            blayer = BLayer.CMaxTable(bigdl_type="float")
+        # error in dot
+        elif klayer.mode == "dot":
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py]: add TODO and throw exception here?

File: pyspark/bigdl/keras1/converter.py [hunk 9]
```diff
@@ -403,16 +408,218 @@ def create_reshape(self, klayer, kclayer):
...
         return blayer
 
+    def create_repeatvector(self, klayer, kclayer):
+        return BLayer.Replicate(n_features=klayer.n,
+                                n_dim=1,
+                                bigdl_type="float")
+
     def create_merge(self, klayer, kclayer):
         self.__check_is_share_weights(kclayer)
+        input_shape = klayer.get_input_shape_at(0)
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py]: add exception if those parameters are not None: `output_shape=None, output_mask=None,
                 arguments=None, node_indices=None, tensor_indices=None`

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 4 above)
Comment (zhichao-li) [reply pyspark/test/bigdl/keras1/test_layer.py]: add unit test for setting diff activation

File: pyspark/bigdl/keras1/converter.py [hunk 10]
```diff
@@ -403,16 +408,218 @@ def create_reshape(self, klayer, kclayer):
...
+        return BLayer.SpatialZeroPadding(0, 0, -cropping[0], -cropping[1])
+
+    # Recurrent Layers
+    def create_simplernn(self, klayer, kclayer):
+        rec = BLayer.Recurrent()
+        input_shape = klayer.get_input_shape_at(0)
+        config = kclayer["config"]
+        activation = self.to_bigdl_activation(config["activation"],
+                                              "%s_%s" % (config["name"], config["activation"]))
+        print(input_shape[2])
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py]: remove this ?

**Comment by zhichao-li** [comment]
invoke `combo_parameter_layer` for layer with parameters

File: pyspark/bigdl/keras1/converter.py [hunk 11]
```diff
@@ -403,16 +408,218 @@ def create_reshape(self, klayer, kclayer):
...
+        print(klayer.output_dim)
+        rnn = BLayer.RnnCell(input_size=input_shape[2],
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py]: how to handle `init='glorot_uniform', inner_init='orthogonal',` and `dropout_W=0., dropout_U=0.` ?

File: pyspark/test/bigdl/keras1/test_layer.py [hunk 12]
```diff
@@ -267,6 +268,120 @@ def test_merge_concat(self):
...
+        inputLayer3 = InputLayer(input_shape=(3, 6, 7))
+
+        layer = Merge([inputLayer1, inputLayer2, inputLayer3], mode='max')
+        input_data = [np.random.random_sample([2, 3, 6, 7]),
+                      np.random.random([2, 3, 6, 7]),
+                      np.random.random([2, 3, 6, 7])]
+        self.__modelTestSingleLayer(input_data, layer, functional_api=False)
+
+    def test_merge_dot(self):
+        # ERROR: different output shapes
```
Comment (zhichao-li) [inline pyspark/test/bigdl/keras1/test_layer.py]: have we solved this problem?

File: pyspark/bigdl/keras1/converter.py [hunk 13]
```diff
@@ -403,16 +408,218 @@ def create_reshape(self, klayer, kclayer):
...
+    def __generate_zeropadding1d(self, pad_top, pad_bottom):
+        return BLayer.SpatialZeroPadding(pad_left=0,
+                                         pad_right=0,
+                                         pad_top=pad_top,
+                                         pad_bottom=pad_bottom,
+                                         bigdl_type="float")
+
+    def create_zeropadding1d(self, klayer, kclayer):
+        padding = klayer.padding
+        if isinstance(padding, int):  # int
```
Comment (zhichao-li) [inline pyspark/bigdl/keras1/converter.py:534]: try to read those padding value from kclayer to get rid of those checking

Comment (zhichao-li) [reply pyspark/bigdl/keras1/converter.py:534]: add unit test for those diff scenario.

File: pyspark/bigdl/keras1/converter.py (hunk 1 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py]: Do we need to check if config["W_constraint"]=None?

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 12 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras1/test_layer.py]: Not yet

File: pyspark/bigdl/keras1/converter.py (hunk 13 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py:534]: I have already added unit test for each different input

Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py:534]: The checking I do here is similar to the code in keras. You mean I can directly get the padding value from kclayer? What type?

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 6 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras1/test_layer.py]: ??

File: pyspark/bigdl/keras1/converter.py (hunk 8 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py]: Dot has some error; cos and ave seemingly not supported in BigDL

File: pyspark/bigdl/keras1/converter.py (hunk 3 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py]: DONE

File: pyspark/bigdl/keras1/converter.py (hunk 2 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py]: DONE

File: pyspark/bigdl/keras1/converter.py (hunk 1 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py]: DONE

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 5 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras1/test_layer.py]: DONE

File: pyspark/bigdl/keras1/converter.py (hunk 10 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py]: DONE

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 4 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras1/test_layer.py]: DONE

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 6 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras1/test_layer.py]: You mean add the following?
         ` def without_batch(batch_shape):`
                   `return batch_shape.shape[1:]`

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 7 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras1/test_layer.py]: DONE

File: pyspark/bigdl/keras1/converter.py (hunk 8 above)
Comment (hkvision) [reply pyspark/bigdl/keras1/converter.py]: Throw exception for dot for now. Fix it later.

File: pyspark/test/bigdl/keras1/test_layer.py (hunk 6 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras1/test_layer.py]: Ignored.
